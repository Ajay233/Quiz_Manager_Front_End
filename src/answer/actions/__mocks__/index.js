export const getAnswers = jest.fn()

export const setCurrentAnswer = jest.fn()

export const deleteAnswer = jest.fn()

export const addAnswer = jest.fn(() => { return { type: "ADD_ANSWER", payload: {} } })

export const updateAnswer = jest.fn()
