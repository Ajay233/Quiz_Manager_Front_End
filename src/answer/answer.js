import React from 'react'
import { Link } from 'react-router-dom'
import history from '../history'

export const Answer = (props) => {

  const handleEdit = () => {
    const { setNotification, setCurrentAnswer, answer } = props
    setNotification()
    setCurrentAnswer(answer);
  }

  const handleDelete = () => {
    const { showModal, setCurrentAnswer, answer } = props
    showModal();
    setCurrentAnswer(answer);
  }

  const renderAnswer = () => {
    const { answerIndex, description, correctAnswer } = props.answer
    return(
      <div className="answerContainer">
        <div className="ansNumber">{ answerIndex }</div>
        <div className="ansDecription">{ description }</div>
        <div className="correctAns">{ renderCorrectAnswer(correctAnswer) }</div>
        {renderOptions()}
      </div>
    );
  }

  const renderCorrectAnswer = (correctAnswer) => {
    if(history.location.pathname !== '/editQuestion'){
      return <i className="fas fa-eye-slash"></i>
    } else {
      return correctAnswer ? <i className="far fa-check-circle green"></i> : <i className="far fa-times-circle red"></i>
    }
  }

  const renderOptions = () => {
    const { permission } = props.userData;
    return(
      <div className="ansOptions">
        { renderDelete(permission) }
        { renderEdit(permission) }
      </div>
    );
  }

  const renderDelete = (permission) => {
    if(history.location.pathname === '/editQuestion'){
      return(
        <Link
          to="#"
          id="deleteAnswer"
          className="deleteOption linkStandard"
          onClick={() => {handleDelete()}}
        >
          <i className="fas fa-trash-alt red"></i> Delete
        </Link>
      );
    } else {
      return null
    }
  }

  const renderEdit = (permission) => {
    if(history.location.pathname === '/editQuestion'){
      return(
        <Link
          to="/editAnswer"
          id="editAnswer"
          className="edit linkStandard"
          onClick={() => {handleEdit()}}
        >
          <i className="fas fa-edit blue"></i> Edit
        </Link>
      );
    } else {
      return null
    }
  }

  return(
    <div>
      {renderAnswer()}
    </div>
  );
}

export default Answer
