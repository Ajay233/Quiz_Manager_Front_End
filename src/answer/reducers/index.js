import { answers } from '../factory/answersFactory'

export const setAnswersReducer = (state=answers, action) => {
  switch (action.type) {
    case "SET_ANSWERS": return action.payload;
    case "DELETE_ANSWER": return state.filter(answer => answer != action.payload)
    default: return state
  }
}


export const setCurrentAnswerReducer = (state={}, action) => {
  switch (action.type) {
    case "SET_CURRENT_ANSWER": return action.payload;
    default: return state;
  }
}
