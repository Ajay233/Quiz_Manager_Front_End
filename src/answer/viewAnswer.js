import React from 'react';
import { connect } from 'react-redux'
import UpdateAnswerForm from '../forms/updateAnswer'
import NewAnswerForm from '../forms/newAnswer'
import Notification from '../notifications/notifications'
import history from '../history'

export class AnswerView extends React.Component {

  componentDidMount(){
    document.documentElement.scrollTop = 0;
  }

  renderForm = () => {
    return history.location.pathname === "/newAnswer" ? <NewAnswerForm /> : <UpdateAnswerForm />;
  }

  render(){
    return(
      <div className="componentContainer">
        <div>
          <Notification />
        </div>
        {this.renderForm()}
      </div>
    );
  }
}

export const mapStateToProps = (state) => {
  return {
    userData: state.userData,
    currentAnswer: state.currentAnswer
  }
}

export default connect(mapStateToProps)(AnswerView)
