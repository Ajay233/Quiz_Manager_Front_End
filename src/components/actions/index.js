export const setToTopButton = (show=false) => {
  return {
    type: "SHOW_TO_TOP_BUTTON",
    payload: show
  }
}

export const setUrl = (url="") => {
  return {
    type: "SET_URL",
    payload: url
  }
}

export const enableLightMode = () => {
  return {
    type: "SET_THEME",
    payload: false
  }
}

export const enableDarkMode = () => {
  return {
    type: "SET_THEME",
    payload: true
  }
}

export const setLoaderState = (show=false, message="", label="") => {
  return {
    type: "SET_LOADER_STATE",
    payload: {
      show: show,
      message: message,
      label: label
    }
  }
}
