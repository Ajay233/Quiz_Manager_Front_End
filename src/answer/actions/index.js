import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { sessionExpired } from '../../utils/session'
import history from '../../history'

// create
export const addAnswer = (data, jwt) => {
  return (dispatch) => {
    return post("answer/create", [data], jwt).then((response) => {
      dispatch({
        type: "ADD_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer created", "success", true));
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create answer", "error", true));
      }
    });
  }
}

// read
export const getAnswers = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("answer/findByQuestionId", param, jwt).then((response) => {
      dispatch({
        type: "SET_ANSWERS",
        payload: response.data
      });
    }).catch((error) => {
      dispatch(setNotification("Error retrieving answers", "error", true));
    });
  }
}

// update (there was a call to setCurrentAnswer but have removed this as it's not needed)
export const updateAnswer = (body, jwt) => {
  return (dispatch) => {
    return put("answer/update", [body], jwt).then((response) => {
      dispatch({
        type: "UPDATE_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer updated", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to update answer", "error", true));
      }
    });
  }
}

// delete
export const deleteAnswer = (config, jwt) => {
  return (dispatch) => {
    return del("answer/delete", config, jwt).then((response) => {
      dispatch(hideModal())
      dispatch({
        type: "DELETE_ANSWER",
        payload: config.data[0]
      })
      dispatch(setNotification("Answer deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(hideModal())
        dispatch(setNotification("Error - Unable to delete this answer", "error", true))
      }
    });
  }
}

export const setCurrentAnswer = (answer) => {
  return {
    type: "SET_CURRENT_ANSWER",
    payload: answer
  }
}
