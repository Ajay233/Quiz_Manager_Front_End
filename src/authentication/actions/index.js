import { post } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import history from '../../history'

export const setUser = (endpoint, loginDetails) => {

  return (dispatch) => {

    post(endpoint, loginDetails).then((response) => {
      dispatch({
        type: "SET_USER_LOGGED_IN",
        payload: {
          id: response.data.user.id,
          forename: response.data.user.forename,
          surname: response.data.user.surname,
          email: response.data.user.email,
          permission: response.data.user.permission,
          verified: response.data.user.verified,
          jwt: response.data.jwt,
          loggedIn: true,
          response: response
        }
      })
      history.push("/")
      dispatch(setNotification(`Welcome back ${response.data.user.forename}`, "loginSuccess", true))
    }).catch((error) => {
      if(error.response.data === "NOT VERIFIED"){
        const verifyMsg = "Your email has not yet been veirified.  You will need to verify your email before you can log in"
        dispatch(setNotification(verifyMsg, "warning", true))
      } else {
        dispatch({
          type: "SET_USER_LOGGED_IN",
          payload: {
            error: error.response
          }
        });
        dispatch(setNotification("The username or password you entered was incorrect", "error", true));
      }
    })
  }
}

export const logOutUser = () => {
  return {
    type: "LOG_OUT_USER",
    payload: {
      forename: "",
      surname: "",
      email: "",
      permission: "",
      verified: "",
      jwt: "",
      loggedIn: false
    }
  }
}

export const setVerficationProcess = (endpoint, verificationDetails) => {

  return (dispatch) => {

    post(endpoint, verificationDetails).then((response) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "completed",
          token: "",
          error: {}
        }
      })
    }).catch((error) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "not verified",
          token: verificationDetails.token,
          error: error.response
        }
      })
    })
  }
}

export const setRedirect = (status) => {
  return {
    type: "SET_REDIRECT",
    payload: {
      status: status
    }
  }
}

export const logOut = () => {
  return {
    type: "RESET_APP"
  }
}

// export const getUsers = (endpoint, header) => {
//   return {
//     type: "GET_USER",
//     payload: {
//       userId
//   }
// }
