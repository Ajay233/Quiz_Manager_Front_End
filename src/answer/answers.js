import React from 'react';
import { Answer } from './answer'

export const Answers = (props) => {

  const renderAnswers = () => {
    const { answers, userData, setCurrentAnswer, setNotification, deleteAnswer, showModal, quiz } = props;
    const listOfAnswers = answers.map(answer => {
      return(
        <Answer
          key={answers.indexOf(answer)}
          answer={answer}
          userData={userData}
          setCurrentAnswer={setCurrentAnswer}
          setNotification={setNotification}
          deleteAnswer={deleteAnswer}
          showModal={showModal}
          quiz={quiz}
        />
      );
    });
    return listOfAnswers;
  }

  return(
    <div>
      {renderAnswers()}
    </div>
  );
}

export default Answers
