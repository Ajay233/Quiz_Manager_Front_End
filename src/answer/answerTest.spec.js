import React from 'react'
import { Router } from 'react-router-dom'
import { Answer } from './answer'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { setCurrentAnswer, deleteAnswer } from './actions'
import { setNotification } from '../notifications/actions'
import { showModal } from '../modal/actions'
import history from '../history'


jest.mock("./actions")
jest.mock("../notifications/actions")
jest.mock("../modal/actions")

describe("Answer", () => {

  let answer;

  beforeEach(() => {
    answer = {
      id: 1,
      answerNumber: 1,
      description: "testDescription",
      correctAnswer: false
    }
  })

  it("should match snapshot for user with 'READ-ONLY' permission", () => {

    const userData = {
      permission: "READ-ONLY",
      jwt: "jwt"
    }

    const component = renderer.create(
      <Answer
        answer={answer}
        userData={userData}
      />
    )

    expect(component).toMatchSnapshot()
  })


  it("should match snapshot for user with 'ADMIN' permission", () => {

    const userData = {
      permission: "ADMIN",
      jwt: "jwt"
    }

    const component = renderer.create(
      <Router history={history}>
        <Answer
          answer={answer}
          userData={userData}
        />
      </Router>
    )

    expect(component).toMatchSnapshot()
  })

  it("should match snapshot which renders a correct answer", () => {

    const correctAnswer = {
      id: 1,
      answerNumber: 1,
      description: "testDescription",
      correctAnswer: true
    }

    const userData = {
      permission: "ADMIN",
      jwt: "jwt"
    }

    const component = renderer.create(
      <Router history={history}>
        <Answer
          answer={correctAnswer}
          userData={userData}
        />
      </Router>
    )

    expect(component).toMatchSnapshot()
  })
})

describe("handleEdit", () => {

  let answer;

  beforeEach(() => {
    answer = {
      id: 1,
      answerNumber: 1,
      description: "testDescription",
      correct: false
    }
  })

  it("should call setCurrentAnswer", () => {

    const userData = {
      permission: "USER",
      jwt: "jwt"
    }

    const wrapper = mount(
      <Answer
        answer={answer}
        userData={userData}
        setCurrentAnswer={setCurrentAnswer}
      />
    )

    wrapper.instance().handleEdit()
    expect(setCurrentAnswer).toHaveBeenCalledTimes(1)
  })
})

describe("handleDelete", () => {

  let answer;

  beforeEach(() => {
    answer = {
      id: 1,
      answerNumber: 1,
      description: "testDescription",
      correct: false
    }
  })

  it("should call showModal and setCurrentAnswer", () => {

    setCurrentAnswer.mockClear()

    const userData = {
      permission: "USER",
      jwt: "jwt"
    }

    const wrapper = mount(
      <Answer
        answer={answer}
        userData={userData}
        setCurrentAnswer={setCurrentAnswer}
        showModal={showModal}
      />
    )

    wrapper.instance().handleDelete()
    expect(setCurrentAnswer).toHaveBeenCalledTimes(1)
    expect(showModal).toHaveBeenCalledTimes(1)

  })
})
