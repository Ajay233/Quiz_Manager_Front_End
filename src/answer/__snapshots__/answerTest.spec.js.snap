// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer should match snapshot for user with 'ADMIN' permission 1`] = `
<div>
  <div
    className="answerContainer"
  >
    <div
      className="ansNumber"
    >
      1
    </div>
    <div
      className="ansDecription"
    >
      testDescription
    </div>
    <div
      className="correctAns"
    >
      <i
        className="fas fa-eye-slash"
      />
    </div>
    <div
      className="ansOptions"
    />
  </div>
</div>
`;

exports[`Answer should match snapshot for user with 'READ-ONLY' permission 1`] = `
<div>
  <div
    className="answerContainer"
  >
    <div
      className="ansNumber"
    >
      1
    </div>
    <div
      className="ansDecription"
    >
      testDescription
    </div>
    <div
      className="correctAns"
    >
      <i
        className="fas fa-eye-slash"
      />
    </div>
    <div
      className="ansOptions"
    />
  </div>
</div>
`;

exports[`Answer should match snapshot which renders a correct answer 1`] = `
<div>
  <div
    className="answerContainer"
  >
    <div
      className="ansNumber"
    >
      1
    </div>
    <div
      className="ansDecription"
    >
      testDescription
    </div>
    <div
      className="correctAns"
    >
      <i
        className="fas fa-eye-slash"
      />
    </div>
    <div
      className="ansOptions"
    />
  </div>
</div>
`;
